sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsNamespaces.java

```diff
@@ -136,15 +137,20 @@ void alterNamespace(
       NamespaceChange... changes) throws NoSuchNamespaceException;
 
   /**
-   * Drop a namespace from the catalog, recursively dropping all objects within the namespace.
+   * Drop a namespace from the catalog with cascade mode, recursively dropping all objects
+   * within the namespace if cascade is true.
    * <p>
    * If the catalog implementation does not support this operation, it may throw
    * {@link UnsupportedOperationException}.
    *
    * @param namespace a multi-part namespace
+   * @param cascade When true, deletes all objects under the namespace
    * @return true if the namespace was dropped
    * @throws NoSuchNamespaceException If the namespace does not exist (optional)
+   * @throws NonEmptyNamespaceException If the namespace is non-empty and cascade is false
    * @throws UnsupportedOperationException If drop is not a supported operation
    */
-  boolean dropNamespace(String[] namespace) throws NoSuchNamespaceException;
+  boolean dropNamespace(
+      String[] namespace,
+      boolean cascade) throws NoSuchNamespaceException, NonEmptyNamespaceException;
```

Comment:
Wait .. this API was added in 3.0.0 at SPARK-27661. Shouldn't we have overwritten version to keep the backward compatibility?

Reply:
Do you mean we should keep the old API `dropNamespace(String[] namespace)`, and add a new API `dropNamespace( String[] namespace, boolean cascade)`? @HyukjinKwon

Reply:
Yes, otherwise the downstream datasources complied against Spark 3.2 would not work with Spark 3.3 with an exception like method does not exist. Mima plugin we use usually detects such binary compatibility but seems it didn't for some reasons.

Reply:
Should I create a PR for reversing this API `dropNamespace` and adding a new one? WDYT? @cloud-fan @HyukjinKwon

Reply:
We changed the implementation of `DropNamespaceExec` to not call `listTables` at the beginning, and the catalog implementation must deal with the `cascade` parameter properly. It's better to break binary compatibility here, instead of silently calling the old `dropNamespace` implementation which may drop all the tables in the namespace.

Reply:
Can we keep the binary compatibility but change the old behavior?

Reply:
I don't think it's possible, unless we add versioning to the DS v2 APIs, so that we know which version the catalog implementation is.

Reply:
Can we implement default method of `dropNamespace(String[] namespace)` that calls `dropNamespace(namespace, cascade = false)` or `dropNamespace(namespace, cascade = true)` with deprecating and pointing the alternative? It already used to work with cascading before, and that's what we documented. This API was added from Spark 3.0.0, and I don't think we should break this binary compatibility unless we must.

Reply:
If we think that we shouldn't make this implemented, we should throw an exception with keeping the signature so we can show nicer error message instead of method not found.

Reply:
> Can we implement default method of dropNamespace(String[] namespace)

This doesn't help,  because the new `dropNamespace(String[] namespace, boolean cascade)` is not implemented and will break all the existing implementations.

A common way to keep backward compatibility is to add default implementation for the new API. e.g.
```
boolean dropNamespace(String[] namespace, boolean cascade) {
  if (cascade) dropNamespace(namespace) else throw ...
}
```
However, `DROP DATABASE` is much more commonly used than `DROP DATABASE ... CASCADE`. So this doesn't help either.

Reply:
Can we at least keep the old signature, deprecate it and throw an exception? Also this PR changed the behaviour.

I know that DSv2 has been less strict on this but probably we should start taking care about this given that the API has been almost 2 years since Spark 3.0.0.